### PR TITLE
[TRUS-3805] - Navigation, Print Helper & Extractor 

### PR DESCRIPTION
**Pull request by jassalrichy**

### CODE REVIEW

File: app/extractors/IndividualBeneficiaryExtractor.scala [hunk 1]
```diff
@@ -82,34 +131,42 @@ class IndividualBeneficiaryExtractor @Inject()() {
...
   private def extractShareOfIncome(individualBeneficiary: IndividualBeneficiary,
                                    answers: UserAnswers) : Try[UserAnswers] = {
-    individualBeneficiary.income match {
-      case Some(income) =>
-        answers.set(IncomeDiscretionYesNoPage, false)
-          .flatMap(_.set(IncomePercentagePage, income.toInt))
-      case None =>
-        // Assumption that user answered yes as the share of income is not provided
-        answers.set(IncomeDiscretionYesNoPage, true)
+    if (answers.isTaxable) {
```
Comment (jackheslop96) [inline app/extractors/IndividualBeneficiaryExtractor.scala:134]: There is a BeneficiaryExtractor trait you could extend that has the share of income, address, and country of residence methods defined. You just need to override the page objects. We also have something similar in maintain-a-trust-frontend to refactor the country of residence and country of nationality methods if you want to implement something similar? (https://github.com/hmrc/maintain-a-trust-frontend/blob/7dd2139d9fa5446a479cd1575b126ce66e869fa7/app/mapping/PlaybackExtractor.scala#L67-L108)

File: app/navigation/IndividualBeneficiaryNavigator.scala [hunk 2]
```diff
@@ -35,31 +35,166 @@ class IndividualBeneficiaryNavigator @Inject()() extends Navigator {
...
+    } else {
+      rts.CountryOfResidenceYesNoController.onPageLoad(mode)
+    }
+  }
+
+  private def navigateAwayFromCountryOfResidencyQuestions(ua: UserAnswers, mode: Mode): Call = {
+    (ua.get(NationalInsuranceNumberYesNoPage), ua.isTaxable) match {
+      case (Some(true), _) => rts.MentalCapacityYesNoController.onPageLoad(mode)
+      case (_, true) => rts.AddressYesNoController.onPageLoad(mode)
+      case _ => rts.MentalCapacityYesNoController.onPageLoad(mode)
```
Comment (jackheslop96) [inline app/navigation/IndividualBeneficiaryNavigator.scala:150]: Could this be done in 2 cases? Perhaps
`case (Some(true), _) | (_, false) => rts.MentalCapacityYesNoController.onPageLoad(mode)`
`case _ => rts.AddressYesNoController.onPageLoad(mode)`

File: app/utils/Constants.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package utils
+
+object Constants {
+  final val GB = "GB"
+  final val ES = "ES"
+  final val UNITED_KINGDOM = "United Kingdom"
+  final val SPAIN = "Spain"
+}
```
Comment (jackheslop96) [inline app/utils/Constants.scala:24]: Think there's an existing `Constant` object that already has `GB`. I prefer the name `Constants` though so you can delete the other one and point the existing usages to this one instead

File: app/extractors/IndividualBeneficiaryExtractor.scala (hunk 1 above)
Comment (jassalrichy) [reply app/extractors/IndividualBeneficiaryExtractor.scala]: Updated

File: app/navigation/IndividualBeneficiaryNavigator.scala (hunk 2 above)
Comment (jassalrichy) [reply app/navigation/IndividualBeneficiaryNavigator.scala]: Updated

File: app/utils/Constants.scala (hunk 3 above)
Comment (jassalrichy) [reply app/utils/Constants.scala]: Updated

File: app/extractors/BeneficiaryExtractor.scala [hunk 4]
```diff
@@ -65,6 +84,21 @@ trait BeneficiaryExtractor[T <: Beneficiary] {
...
+      .flatMap(_.set(mentalCapacityYesNoPage, entity.mentalCapacityYesNo))
+      .flatMap(_.set(vulnerableYesNoPage, entity.vulnerableYesNo))
+      .flatMap(answers => extractDateOfBirth(entity.dateOfBirth, answers))
+      .flatMap(answers => extractShareOfIncome(entity.income, answers))
+      .flatMap(answers => extractCountryOfNationality(entity.nationality, answers))
+      .flatMap(answers => extractIdentification(entity.identification, entity.address, answers))
+      .flatMap(answers => extractCountryOfResidence(entity.countryOfResidence, answers))
+      .flatMap(answers => extractAddress(entity.address, answers))
+  }
+
```
Comment (jackheslop96) [inline app/extractors/BeneficiaryExtractor.scala:101]: I would keep this in the `IndividualBeneficiaryExtractor` as it's only specific to that

File: app/extractors/BeneficiaryExtractor.scala [hunk 5]
```diff
@@ -79,6 +113,36 @@ trait BeneficiaryExtractor[T <: Beneficiary] {
...
+  def extractDateOfBirth(dateOfBirth: Option[LocalDate], answers: UserAnswers) : Try[UserAnswers] = {
+    dateOfBirth match {
+      case Some(dob) =>
+        answers.set(dateOfBirthYesNoPage, true)
+          .flatMap(_.set(dateOfBirthPage, dob))
+      case None =>
+        // Assumption that user answered no as dob is not provided
+        answers.set(dateOfBirthYesNoPage, false)
+    }
+  }
```
Comment (jackheslop96) [inline app/extractors/BeneficiaryExtractor.scala:125]: Likewise I would keep this in the `IndividualBeneficiaryExtractor`

File: app/extractors/BeneficiaryExtractor.scala [hunk 6]
```diff
@@ -117,4 +181,42 @@ trait BeneficiaryExtractor[T <: Beneficiary] {
...
+
+  def extractPassportOrIdCardDetailsYesNo(hasAddress: Boolean, answers: UserAnswers): Try[UserAnswers] = {
+    if (answers.isTaxable && hasAddress) {
+      answers.set(passportOrIdCardDetailsYesNoPage, false)
+    } else {
+      Success(answers)
+    }
+  }
```
Comment (jackheslop96) [inline app/extractors/BeneficiaryExtractor.scala:219]: Likewise I would keep this in the `IndividualBeneficiaryExtractor`

File: app/extractors/BeneficiaryExtractor.scala [hunk 7]
```diff
@@ -117,4 +181,42 @@ trait BeneficiaryExtractor[T <: Beneficiary] {
...
+            .flatMap(_.set(passportOrIdCardDetailsYesNoPage, true))
+            .flatMap(_.set(passportOrIdCardDetailsPage, combined))
+        case _ =>
+          answers.set(nationalInsuranceNumberYesNoPage, false)
+            .flatMap(answers => extractPassportOrIdCardDetailsYesNo(address.isDefined, answers))
+      }
+    } else {
+      Success(answers)
+    }
+  }
```
Comment (jackheslop96) [inline app/extractors/BeneficiaryExtractor.scala:211]: Likewise I would keep this in the `IndividualBeneficiaryExtractor`

File: app/models/beneficiaries/Beneficiary.scala [hunk 8]
```diff
@@ -33,6 +32,19 @@ trait OrgBeneficiary extends Beneficiary {
...
+  val dateOfBirth: Option[LocalDate]
+  val roleInCompany: Option[RoleInCompany]
+  val income: Option[String]
+  val nationality: Option[String]
+  val countryOfResidence: Option[String]
+  val identification: Option[IndividualIdentification]
+  val address: Option[Address]
+  val mentalCapacityYesNo: Option[Boolean]
+  val vulnerableYesNo: Option[Boolean]
+}
```
Comment (jackheslop96) [inline app/models/beneficiaries/Beneficiary.scala:46]: Don't think this is needed

File: app/extractors/BeneficiaryExtractor.scala (hunk 4 above)
Comment (jassalrichy) [reply app/extractors/BeneficiaryExtractor.scala]: Updated

File: app/extractors/BeneficiaryExtractor.scala (hunk 5 above)
Comment (jassalrichy) [reply app/extractors/BeneficiaryExtractor.scala]: Updated

File: app/extractors/BeneficiaryExtractor.scala (hunk 6 above)
Comment (jassalrichy) [reply app/extractors/BeneficiaryExtractor.scala]: Updated

File: app/extractors/BeneficiaryExtractor.scala (hunk 7 above)
Comment (jassalrichy) [reply app/extractors/BeneficiaryExtractor.scala]: Updated

File: app/models/beneficiaries/Beneficiary.scala (hunk 8 above)
Comment (jassalrichy) [reply app/models/beneficiaries/Beneficiary.scala]: Removed

**Review by jackheslop96 — APPROVED** [review]

**Pull request merged**